File: objective-c-sample/SendBirdSampleUI/SendBird/SendBirdChat/View/ChannelListTableView.m
//
//  ChannelListTableView.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#import "ChannelListTableView.h"

@interface ChannelListTableView ()<UITableViewDataSource, UITableViewDelegate, UITextFieldDelegate> {
    SendBirdChannelListQuery *channelListQuery;
}

@end

@implementation ChannelListTableView

- (id) initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self != nil) {
        [self initView];
    }
    return self;
}

- (void) initView
{
    [self setBackgroundColor:[UIColor whiteColor]];
    self.searchAreaView = [[UIView alloc] init];
    [self.searchAreaView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.searchAreaView setBackgroundColor:UIColorFromRGB(0x2ac6b6)];
    self.searchTextField = [[UITextField alloc] init];
    [self.searchTextField setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.searchTextField setBackground:[UIImage imageNamed:@"_box_white"]];
    [self.searchTextField setTextColor:[UIColor blackColor]];
    [self.searchTextField setFont:[UIFont systemFontOfSize:13.0]];
    UIView *leftPaddingView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 16, 8)];
    UIView *rightPaddingView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 16, 8)];
    [self.searchTextField setLeftView:leftPaddingView];
    [self.searchTextField setRightView:rightPaddingView];
    [self.searchTextField setLeftViewMode:UITextFieldViewModeAlways];
    [self.searchTextField setRightViewMode:UITextFieldViewModeAlways];
    [self.searchTextField setReturnKeyType:UIReturnKeySearch];
    [self.searchTextField setDelegate:self];
    [self.searchTextField addTarget:self action:@selector(searchTextFieldDidChange:) forControlEvents:UIControlEventEditingChanged];
    
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] initWithString:@"  Search"];
    NSTextAttachment *textAttachment = [[NSTextAttachment alloc] init];
    textAttachment.image = [UIImage imageNamed:@"_icon_search"];
    NSAttributedString *attrStringWithImage = [NSAttributedString attributedStringWithAttachment:textAttachment];
    [attributedString replaceCharactersInRange:NSMakeRange(0, 1) withAttributedString:attrStringWithImage];
    [self.searchTextField setAttributedPlaceholder:attributedString];
    
    self.channelTableView = [[UITableView alloc] init];
    [self.channelTableView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.channelTableView setDelegate:self];
    [self.channelTableView setDataSource:self];
    [self.channelTableView setSeparatorColor:[UIColor clearColor]];
    
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self action:@selector(reloadChannels) forControlEvents:UIControlEventValueChanged];
    
    [self.searchAreaView addSubview:self.searchTextField];
    [self addSubview:self.searchAreaView];
    
    [self.channelTableView addSubview:self.refreshControl];
    [self addSubview:self.channelTableView];
    
    [self applyConstraints];
}

- (void)viewDidLoad
{
    channelListQuery = [SendBird queryChannelList];
    self.channels = [[NSMutableArray alloc] init];
    [ImageCache initImageCache];
}

- (void) applyConstraints
{
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.searchAreaView attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTop multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.searchAreaView attribute:NSLayoutAttributeLeading relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeLeading multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.searchAreaView attribute:NSLayoutAttributeTrailing relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTrailing multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.searchAreaView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:43]];
    
    [self.searchAreaView addConstraint:[NSLayoutConstraint constraintWithItem:self.searchTextField attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self.searchAreaView attribute:NSLayoutAttributeCenterY multiplier:1 constant:0]];
    [self.searchAreaView addConstraint:[NSLayoutConstraint constraintWithItem:self.searchTextField attribute:NSLayoutAttributeLeading relatedBy:NSLayoutRelationEqual toItem:self.searchAreaView attribute:NSLayoutAttributeLeading multiplier:1 constant:8]];
    [self.searchAreaView addConstraint:[NSLayoutConstraint constraintWithItem:self.searchTextField attribute:NSLayoutAttributeTrailing relatedBy:NSLayoutRelationEqual toItem:self.searchAreaView attribute:NSLayoutAttributeTrailing multiplier:1 constant:-8]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.searchTextField attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:30]];
    
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.searchAreaView attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self.channelTableView attribute:NSLayoutAttributeTop multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:self.channelTableView attribute:NSLayoutAttributeBottom multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.channelTableView attribute:NSLayoutAttributeLeading relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeLeading multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.channelTableView attribute:NSLayoutAttributeTrailing relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeTrailing multiplier:1 constant:0]];
}

- (void) loadChannels
{
    if ([channelListQuery isLoading]) {
        return;
    }
    
    if (![channelListQuery hasNext]) {
        return;
    }
    
    [channelListQuery nextWithResultBlock:^(NSMutableArray *queryResult) {
        if ([channelListQuery page] == 1) {
            [self.channels removeAllObjects];
        }
        [self.channels addObjectsFromArray:queryResult];
        
        [self.channelTableView reloadData];
        [self.refreshControl endRefreshing];
    } endBlock:^(NSError *error) {
        NSLog(@"Error");
    }];
}

- (void) reloadChannels
{
    //    [[[SendBird sharedInstance] taskQueue] cancelAllOperations];
    [self.chattingTableViewController setIndicatorHidden:NO];
    channelListQuery = [SendBird queryChannelList];
    [channelListQuery nextWithResultBlock:^(NSMutableArray *queryResult) {
        if ([channelListQuery page] == 1) {
            [self.channels removeAllObjects];
        }
        [self.channels addObjectsFromArray:queryResult];
        
        [self.channelTableView reloadData];
        [self.refreshControl endRefreshing];
        
        [self.chattingTableViewController setIndicatorHidden:YES];
    } endBlock:^(NSError *error) {
        NSLog(@"Error");
        [self.chattingTableViewController setIndicatorHidden:YES];
    }];
}

- (void) queryChannel:(NSString *)query
{
    [self.chattingTableViewController setIndicatorHidden:NO];
    channelListQuery = [SendBird queryChannelListWithKeyword:query];
    [channelListQuery nextWithResultBlock:^(NSMutableArray *queryResult) {
        if ([channelListQuery page] == 1) {
            [self.channels removeAllObjects];
        }
        [self.channels addObjectsFromArray:queryResult];
        
        [self.channelTableView reloadData];
        [self.refreshControl endRefreshing];
        
        [self.chattingTableViewController setIndicatorHidden:YES];
    } endBlock:^(NSError *error) {
        [self.chattingTableViewController setIndicatorHidden:YES];
    }];
}

- (void)searchTextFieldDidChange:(UITextField *)textField
{
    if ([[textField text] length] == 0) {
        [self reloadChannels];
    }
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [self.channels count];;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"ChannelReuseIdentifier";
    
    ChannelTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    
    if (cell == nil) {
        cell = [[ChannelTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    [cell setModel:[self.channels objectAtIndex:indexPath.row]];
    
    if (indexPath.row == [self.channels count] - 1 && [channelListQuery hasNext]) {
        [self loadChannels];
    }
    
    return cell;
}

#pragma mark - UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 45;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    SendBirdChannel *channel = [_channels objectAtIndex:indexPath.row];
    [self.chattingTableViewController setChannelUrl:[channel url]];
    [self.chattingTableViewController initChannelTitle];
    [self.chattingTableViewController setViewMode:kChattingViewMode];
    
    [self.chattingTableViewController startChatting];
    [self setHidden:YES];
}

#pragma mark - ChannelListQueryResultHandlerDelegate
- (void) onResult:(NSMutableArray **)queryResult
{
    if ([channelListQuery page] == 1) {
        [self.channels removeAllObjects];
    }
    [self.channels addObjectsFromArray:*queryResult];
    
    [self.channelTableView reloadData];
    [self.refreshControl endRefreshing];
}

- (void) onError
{
    
}

#pragma mark - UITextFieldDelegate
- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    if ([[textField text] length] > 0) {
        [self queryChannel:[textField text]];
    }
    return YES;
}

@end

File: swift-sample/SendBirdSampleUI/SendBird/SendBirdChat/View/ChannelListTableView.swift
//
//  ChannelListTableView.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/2/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

class ChannelListTableView: UIView, UITextFieldDelegate, UITableViewDataSource, UITableViewDelegate {
    var searchAreaView: UIView?
    var searchTextField: UITextField?
    var channelTableView: UITableView?
    var channels: NSMutableArray?
    var messagingChannels: Array<AnyObject>?
    var refreshControl: UIRefreshControl?
    var tableViewDataSource: UITableViewDataSource?
    var tableViewDelegate: UITableViewDelegate?
    var chattingTableViewController: ChattingTableViewController?

    private var channelListQuery: SendBirdChannelListQuery?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        initView()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func initView() {
        self.backgroundColor = UIColor.whiteColor()
        self.searchAreaView = UIView()
        self.searchAreaView?.translatesAutoresizingMaskIntoConstraints = false
        self.searchAreaView?.backgroundColor = SendBirdUtils.UIColorFromRGB(0x2ac6b6)

        self.searchTextField = UITextField()
        self.searchTextField?.translatesAutoresizingMaskIntoConstraints = false
        self.searchTextField?.background = UIImage.init(named: "_box_white")
        self.searchTextField?.textColor = UIColor.blackColor()
        self.searchTextField?.font = UIFont.systemFontOfSize(13.0)
        let leftPaddingView = UIView.init(frame: CGRectMake(0, 0, 16, 8))
        let rightPaddingView = UIView.init(frame: CGRectMake(0, 0, 16, 8))
        self.searchTextField?.leftView = leftPaddingView
        self.searchTextField?.rightView = rightPaddingView
        self.searchTextField?.leftViewMode = UITextFieldViewMode.Always
        self.searchTextField?.rightViewMode = UITextFieldViewMode.Always
        self.searchTextField?.returnKeyType = UIReturnKeyType.Search
        self.searchTextField?.delegate = self
        self.searchTextField?.addTarget(self, action: "searchTextFieldDidChange:", forControlEvents: UIControlEvents.EditingChanged)
        
        let attributedString: NSMutableAttributedString = NSMutableAttributedString.init(string: "  Search")
        let textAttachment: NSTextAttachment = NSTextAttachment()
        textAttachment.image = UIImage.init(named: "_icon_search")
        let attrStringWithImage: NSAttributedString = NSAttributedString.init(attachment: textAttachment)
        attributedString.replaceCharactersInRange(NSMakeRange(0, 1), withAttributedString: attrStringWithImage)
        self.searchTextField?.attributedPlaceholder = attributedString
        
        self.channelTableView = UITableView()
        self.channelTableView?.translatesAutoresizingMaskIntoConstraints = false
        self.channelTableView?.delegate = self
        self.channelTableView?.dataSource = self
        self.channelTableView?.separatorColor = UIColor.clearColor()
        
        self.refreshControl = UIRefreshControl()
        self.refreshControl?.addTarget(self, action: "reloadChannels", forControlEvents: UIControlEvents.ValueChanged)
        
        self.searchAreaView?.addSubview(self.searchTextField!)
        addSubview(self.searchAreaView!)
        
        self.channelTableView?.addSubview(self.refreshControl!)
        addSubview(self.channelTableView!)
        
        self.applyConstraints()
    }
    
    func viewDidLoad() {
        self.channelListQuery = SendBird.queryChannelList()
        self.channels = NSMutableArray()
        ImageCache.initImageCache()
    }
    
    private func applyConstraints() {
        self.addConstraint(NSLayoutConstraint.init(item: self.searchAreaView!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.searchAreaView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.searchAreaView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.searchAreaView!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 43))
        
        self.searchAreaView?.addConstraint(NSLayoutConstraint.init(item: self.searchTextField!, attribute: NSLayoutAttribute.CenterY, relatedBy: NSLayoutRelation.Equal, toItem: self.searchAreaView!, attribute: NSLayoutAttribute.CenterY, multiplier: 1, constant: 0))
        self.searchAreaView?.addConstraint(NSLayoutConstraint.init(item: self.searchTextField!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.searchAreaView!, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 8))
        self.searchAreaView?.addConstraint(NSLayoutConstraint.init(item: self.searchTextField!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.searchAreaView!, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -8))
        self.addConstraint(NSLayoutConstraint.init(item: self.searchTextField!, attribute: NSLayoutAttribute.Height, relatedBy: NSLayoutRelation.Equal, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: 30))

        self.addConstraint(NSLayoutConstraint.init(item: self.searchAreaView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.channelTableView!, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.channelTableView!, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.channelTableView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.channelTableView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: 0))
    }
    
    func loadChannels() {
        if self.channelListQuery?.isLoading() == true {
            return
        }
        
        if self.channelListQuery?.hasNext() == false {
            return
        }
        
        self.channelListQuery?.nextWithResultBlock({ (queryResult) -> Void in
                if self.channelListQuery?.page == 1 {
                    self.channels?.removeAllObjects()
                }
                self.channels?.addObjectsFromArray(queryResult as [AnyObject])
                self.channelTableView?.reloadData()
                self.refreshControl?.endRefreshing()
            }, endBlock: { (error) -> Void in
                NSLog("Error")
        })
    }

    func reloadChannels() {
        self.chattingTableViewController?.setIndicatorHidden(false)
        self.channelListQuery = SendBird.queryChannelList()
        self.channelListQuery?.nextWithResultBlock({ (queryResult) -> Void in
            if self.channelListQuery?.page == 1 {
                self.channels?.removeAllObjects()
            }
            self.channels?.addObjectsFromArray(queryResult as [AnyObject])
            
            self.channelTableView?.reloadData()
            self.refreshControl?.endRefreshing()
            
            self.chattingTableViewController?.setIndicatorHidden(true)
            }, endBlock: { (error) -> Void in
                NSLog("Error")
                self.chattingTableViewController?.setIndicatorHidden(true)
        })
    }
    
    func queryChannel(query: String) {
        self.chattingTableViewController?.setIndicatorHidden(false)
        self.channelListQuery = SendBird.queryChannelListWithKeyword(query)
        self.channelListQuery?.nextWithResultBlock({ (queryResult) -> Void in
            if self.channelListQuery?.page == 1 {
                self.channels?.removeAllObjects()
            }
            self.channels?.addObjectsFromArray(queryResult as [AnyObject])
            
            self.channelTableView?.reloadData()
            self.refreshControl?.endRefreshing()
            
            self.chattingTableViewController?.setIndicatorHidden(true)
            }, endBlock: { (error) -> Void in
                self.chattingTableViewController?.setIndicatorHidden(true)
        })
    }
    
    // MARK: UITableViewDataSource
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return (self.channels?.count)!
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let CellIdentifier: String = "ChannelReuseIdentifier"
        
        var cell: ChannelTableViewCell? = tableView.dequeueReusableCellWithIdentifier(CellIdentifier) as? ChannelTableViewCell
        
        if cell == nil {
            cell = ChannelTableViewCell.init(style: UITableViewCellStyle.Default, reuseIdentifier: CellIdentifier)
        }
        cell!.setModel(self.channels![indexPath.row] as? SendBirdChannel)
        
        if indexPath.row == (self.channels?.count)! - 1 && self.channelListQuery?.hasNext() == true {
            self.loadChannels()
        }
        
        return cell!;
    }
    
    // MARK: UITableViewDelegate
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 45
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let channel: SendBirdChannel = self.channels![indexPath.row] as! SendBirdChannel
        self.chattingTableViewController!.channelUrl = channel.url
        self.chattingTableViewController!.initChannelTitle()
        self.chattingTableViewController!.setViewMode(kChattingViewMode)
        
        self.chattingTableViewController!.startChatting()
        self.hidden = true
    }
    
    // MARK: UITextFieldDelegate
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if textField.text?.characters.count > 0 {
            self.queryChannel(textField.text!)
        }
        return true
    }
}
